### ISSUE TITLE
Possible hoarding of Docker tasks when using replicated-job

### ISSUE THREAD

**Issue by pklejch** [moby/moby#42789]
**Description**

When running Docker service in mode `replicated-job`, after service is removed the underlying Docker tasks are not removed. This leads to hoarding of Docker tasks. I'm not exactly sure if it's a problem, but my gut tells me that if there is a limit, someone will eventually hit it.

**Steps to reproduce the issue:**
1. `docker service create --name test --mode replicated-job --replicas 100 --max-concurrent 25  alpine echo hello`
3. `docker service rm test`
2. From iPython 
```
import docker
client = docker.from_env()

In [57]: len(client.api.tasks(filters={'name': 'test'}))
Out[57]: 100
```
4. `docker service create --name test --mode replicated-job --replicas 100 --max-concurrent 25  alpine echo hello`
3. `docker service rm test`
5. Again from iPython
```
In [58]: len(client.api.tasks(filters={'name': 'test'}))
Out[58]: 200
```
6. And so on...

**Describe the results you received:**
After removing service in mode `replicated-job` the underlying tasks are not removed.

**Describe the results you expected:**
When creating regular Swarm service in `replicated` mode the underlying tasks are removed after removing of the service. I would expect same behavior in `replicated-job` mode.

1. Create regular service `docker service create --name test --mode replicated --replicas 100 alpine sleep infinity`
2. From iPython
```
import docker
client = docker.from_env()

In [114]: len(client.api.tasks(filters={'name': 'test'}))
Out[114]: 100
```
3. Remove the service `docker service rm test`
4. From iPython, number of tasks will eventually converge to zero
```
In [122]: len(client.api.tasks(filters={'name': 'test'}))
Out[122]: 0
```

**Additional information you deem important (e.g. issue happens only occasionally):**
None

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:53:57 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.8
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.6
  Git commit:       75249d8
  Built:            Fri Jul 30 19:52:06 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.9
  GitCommit:        e25210fe30a0a703442421b0f60afac609f950a3
 runc:
  Version:          1.0.1
  GitCommit:        v1.0.1-0-g4144b63
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.6.1-docker)
  scan: Docker Scan (Docker Inc., v0.8.0)

Server:
 Containers: 211
  Running: 6
  Paused: 0
  Stopped: 205
 Images: 419
 Server Version: 20.10.8
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: q9l6i5255gb9xge3b4npoo55c
  Is Manager: true
  ClusterID: yg59vfusbb8hb0vgsjncn49wo
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8  
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 192.168.100.4
  Manager Addresses:
   192.168.100.4:2377
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: e25210fe30a0a703442421b0f60afac609f950a3
 runc version: v1.0.1-0-g4144b63
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.11.0-31-generic
 Operating System: Ubuntu 21.04
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.38GiB
 Name: petr-ntb
 ID: KWOD:WS7T:CO4R:TCXT:WHBI:CGUS:YKND:GRVD:DZD3:IXN5:XYLW:A4NU
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: pklejch
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Lenovo T470p

**Comment by pklejch** [comment]
Well, it really starts to causing slowdowns. On one of our Docker Swarm nodes (80 CPU, 500 GB RAM) just creating simple service takes almost 2.5 minutes
```
$ time docker service create --name test --mode replicated --replicas 3 alpine sleep infinity
lgttovk0ikecsfgdqkcq2pvlv
overall progress: 3 out of 3 tasks 
1/3: running   [==================================================>] 
2/3: running   [==================================================>] 
3/3: running   [==================================================>] 
verify: Service converged 

real	2m26.794s
user	0m1.537s
sys	0m0.460s
```
I'm really suspecting "orphaned" tasks, we currently have ~230k of those.
```
>>> import docker
>>> client = docker.from_env()
>>> len(client.api.tasks())

238359
```

**Comment by thaJeztah** [comment]
@dperny PTAL

**Comment by nblazincic** [comment]
I have same issue here, can we have an update on this ? Thanks

**Comment by pklejch** [comment]
> I have same issue here, can we have an update on this ? Thanks

We have hotfixed it for the time being with simple script in the cron.
1. Using curl and Docker API  we retrieve IDs of completed tasks. https://docs.docker.com/engine/api/v1.41/#tag/Task/operation/TaskList
2. Then using swarmctl binary (https://github.com/moby/swarmkit) we removed those tasks.
```
swarmctl task remove --socket /var/run/docker/swarm/control.sock "$task_id"
```

**Comment by nblazincic** [comment]
Tnx @pklejch . Not really an option for us, we are running db migrations as a CD deployment step in a replicated-job, and we need to be sure that we will have only one task started and executed.

**Comment by Vanav** [comment]
Seems this issue is fixed in Docker 24 (https://github.com/moby/moby/issues/44312).

**Comment by pklejch** [comment]
Fixed in #45106
